### ISSUE TITLE
How are the FIR low-pass filter parameters chosen?

### ISSUE THREAD

**Issue by fumin** [ThomasHabets/sparslog#1]
The current implementation adds a FIR low-pass filter in
https://github.com/ThomasHabets/sparslog/blob/main/src/main.rs#L332

I wonder how are parameters `cutoff=50000` and `twidth=10000` chosen?
Is `cutoff=50000` chosen so that it is slightly larger than the baud rate `baud=38383.5`?

In addition, both [rtl_433](https://github.com/merbanan/rtl_433/blob/2a7fe211cee2c54ef6d47827324176d49a4e942d/src/devices/ikea_sparsnas.c#L76) and [strigeus](https://github.com/strigeus/sparsnas_decoder/blob/master/sparsnas_decode.cpp#L276) don't seem to apply after low-pass filters.
I wonder if you could explain a bit on how this low-pass filter helps in decoding sparsnas' signal?

By the way, thanks for sharing this neatly written library.

**Comment by ThomasHabets** [comment]
Both `cutoff` and `twidth`, as well as the what to downsample to (200ksps) were chosen pretty arbitrarily, and but found to work well.

Yes, `cutoff` of 50k needs to be bigger than the baud rate, and of course also needs to be big enough to actually capture both sides of the FSK signal (don't remember now how separated the ones and zeroes are).

`twidth` could probably be much smaller, and it could potentially improve decode / reduce impact of interference. I originally set it to 10k in order to minimize the number of taps (I implemented FIR filter before I implemented FFT filter. FFT filter speed is not affected by tap count), but didn't change it after I switched to FFT filter.

I was not aware of rtl_433, and I didn't look much at exactly how strigeus does the RF parts, but the main reason I filter is actually because I want to dowsample. Downsampling early means less data to deal with (for any debug output) and better performance on downstream blocks.

I'm not saying my code is more efficient than those other ones. In fact it's very unlikely, since at least strigeus does all of the steps on a per sample basis, with no intermediate "streams". That's going to be more efficient, but also less flexible.

A properly tuned filter (which, as I said, this is not), should probably make for a more reliable decoder than strigeus, though. But it's already good enough for me. I get pretty much 100% decodes in my house.

GNU Radio, and me as well, have some ideas about how to get the best of both worlds: Clear separation of steps, but at the same time compile down to basically processing a sample at a time (or small batch, to utilize SIMD) through the pipeline, instead of having every step do the full batch. But we're not there yet.

**Comment by fumin** [comment]
Thanks Thomas for generously sharing your considerations behind the code. Very helpful for me!

Regarding the balance between maximizing the number of taps versus computational efficiency, I just discovered that what [strigeus](https://github.com/strigeus/sparsnas_decoder/blob/master/sparsnas_decode.cpp#L259) is doing is actually an ingenious way of setting the number of taps to infinity with O(1) computation!
Precisely, it seems that [strigeus](https://github.com/strigeus/sparsnas_decoder/blob/master/sparsnas_decode.cpp#L265C7-L265C12) is computing
```math
\left| \int_{-\infty }^{\infty }H\left( -t-\tau \right)X\left( t \right)e^{i\omega t} - \int_{-\infty }^{\infty }H\left( -t-(\tau-baudrate) \right)X\left( t \right)e^{i\omega (t)} \right|^{2}
```
where $\tau$ is the current time, $H$ the Heaviside function, $X$ the signal, and $\omega$ the frequency.
In other words, he is computing the difference between the fourier coefficient of $\omega$ at time $\tau$ and $\tau-baudrate$.
If at time $\tau$, the fourier coefficient difference of $\omega_{1}$ is smaller than that of $\omega_{2}$, that means the frequency at time $\tau$ has moved towards $\omega_{1}$.

I think the drawback of this approach is that it does allow any drift in the $baudrate$, which you handle nicely in [ZeroCrossing](https://github.com/ThomasHabets/rustradio/blob/main/src/symbol_sync.rs).
Thanks again for this library and for sharing your high level considerations behind the code.

**Comment by ThomasHabets** [comment]
Interesting. Thanks for the explanation. I think you're ahead of me on signal theory, so I'll need to think about that more.

Ideally ZeroCrossing should frequency lock, zeroing in the frequency offset that the packet seems to have during preamble. In practice the sender clock (baud rate) seems accurate enough, and the packets short enough to not have a noticeable drift.

I want to implement [something like this](https://youtu.be/uMEfx_l5Oxk), really, which seems like more the general case.

**Comment by fumin** [comment]
Thanks for sharing your experience with ZeroCrossing block.

The lecture by Andy Walls is really deep, I have to admit it's also above pay grade.
I'd be following your work closely should you happen to implement the Timing Error Detector. : )

Closing this issue for now, and thanks again for sharing.
